app/services/lifetime_usages/flag_refresh_from_invoice_service.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+module LifetimeUsages
+  class FlagRefreshFromInvoiceService < BaseService
+    def initialize(invoice:)
+      @invoice = invoice
+      super
+    end
+
+    def call
+      return result unless invoice.subscription?
+      return result unless invoice.finalized? || invoice.voided?
+      return result unless has_plan_usage_thresholds?
+
+      result.lifecycle_usages = []
+
+      invoice.subscriptions.each do |subscription|
+        lifetime_usage = subscription.lifetime_usage
+        lifetime_usage ||= subscription.build_lifetime_usage(organization: subscription.organization)
+        lifetime_usage.recalculate_invoiced_usage = true
+        lifetime_usage.save!
+
+        result.lifecycle_usages << lifetime_usage
+      end
+
+      result
+    rescue ActiveRecord::RecordInvalid => e
+      result.record_validation_failure!(record: e.record)
+    end
+
+    private
+
+    attr_reader :invoice
+
+    def has_plan_usage_thresholds?
+      invoice.subscriptions.any? { |s| s.plan.usage_thresholds.any? }
```

Comment:
not sure if it really matters, but doing it like this might also lead to unnecessary work performed for all subscriptions if one of them has thresholds. But this is likely premature optimization, so also happy to keep as is.